cts/scheduler/summary/pending-node-no-uname.summary

```diff
@@ -1,26 +1,23 @@
 Using the original execution date of: 2023-02-21 12:19:57Z
 Current cluster status:
   * Node List:
-    * Node node-2: UNCLEAN (offline)
+    * Node node-2: pending
     * Online: [ node-1 ]
 
   * Full List of Resources:
     * st-sbd	(stonith:external/sbd):	 Stopped
 
 Transition Summary:
-  * Fence (reboot) node-2 'node is unclean'
-  * Start      st-sbd     ( node-1 )
+  * Start      st-sbd     ( node-1 )  blocked
```

Comment:
I'm a little surprised the start on the other node gets blocked, but it makes sense since probes will be blocked on the pending node, and without fencing it, we can't safely do anything.

However that does mean that if a node gets "stuck" as pending, the cluster will be blocked. I'm not sure how relevant that is in the real world, but it makes me a little uneasy. Do we have some sort of timer on the controller side for pending nodes?

Reply:
That sounds reasonable. What should we do if it pops?

Reply:
Good question. I've also been thinking about this. Probably issuing a fencing targeting the pending node will be more straight-forward and safe.

Users might prefer a graceful shutdown of the peer node though, but I'm not sure it's possible or sensible at all to request a shutdown remotely via corosync.

Another option could be ignore the pending node and proceed resource management. If this also makes sense, we could probably makes it configurable. Then another question will be, which should be the default behavior?

In any case, there needs to be a way to tell scheduler the state about "pending timed out", for example a special value/combination of the `join` and/or `expected` attribute for `node_state`?  Another question is, does it make sense to differentiate the status  from the normal  "pending" shown from UIs?

Reply:
Maybe:
* Bump feature set
* If DC feature set is compatible, record in_ccm and crmd as the timestamp of when the node joined instead of boolean. 0 indicates not joined.
* If in_ccm is nonzero and crmd is 0 and expected is member, the scheduler can compare the in_ccm time against the pending node timeout, and take whatever action. (Scheduler must accept boolean values for backward compatibility with saved CIBs, and for rolling upgrades when the DC hasn't been updated yet.)

Fencing would be safest. I'd be wary of assuming a joining node has no active resources.

Reply:
Of course if stonith-enabled=false, it's trickier. I guess in that case we could proceed as if expected=down (after the timeout). Alternatively maybe we could nack the node join.

Reply:
> Maybe:
> 
> * Bump feature set
> * If DC feature set is compatible, record in_ccm and crmd as the timestamp of when the node joined instead of boolean. 0 indicates not joined.

Good idea.

> * If in_ccm is nonzero and crmd is 0 and expected is member, 

In this case when a node is pending on joining CPG, there's no `expected` field yet, which is considered to be expected=down.

> the scheduler can compare the in_ccm time against the pending node timeout, and take whatever action. (Scheduler must accept boolean values for backward compatibility with saved CIBs, and for rolling upgrades when the DC hasn't been updated yet.)
> 
> Fencing would be safest. I'd be wary of assuming a joining node has no active resources.

Ok, the current implementation issues fencing upon timeout.

Reply:
> Of course if stonith-enabled=false, it's trickier. I guess in that case we could proceed as if expected=down (after the timeout). Alternatively maybe we could nack the node join.

For the case of stonith-enabled=false, so far there's no logic/concept about node being pending at all. Nothing really blocks resource management as long as it's quorate. I'm not sure if it makes sense to change the behavior.

Reply:
@gao-yan ,

Revisiting this, I'm not sure what the ideal behavior is. It's possible (and has legitimate use cases) for someone to start Corosync on a node but not Pacemaker. It would join the cluster (and contribute to quorum) but not join CPG. On the one hand, blocking resource management on other nodes in this situation is undesirable, but on the other, we can't manage resources if we can't probe an active node.

I'm thinking of reverting the node-pending-timeout portion of this PR. Alternatively maybe we could raise the default to something high (maybe 3h?).

Before node-pending-timeout, would it have been possible for someone to use a permanently Corosync-only node as a quorum node, or would it have blocked resource management? If so, I'd lean to reverting (at least until the next major-version release) to not break that use case. If not, a high default may be sufficient for other use cases (what comes to mind is someone having cluster trouble starting Corosync manually to make sure it comes up OK and maybe do some troubleshooting before starting Pacemaker manually).

Reply:
FYI, I'd revert "add pending-node-timeout" through "trigger a new transition". I think everything else would be useful to keep.

Reply:
> @gao-yan ,
> 
> Revisiting this, I'm not sure what the ideal behavior is. It's possible (and has legitimate use cases) for someone to start Corosync on a node but not Pacemaker. It would join the cluster (and contribute to quorum) but not join CPG. On the one hand, blocking resource management on other nodes in this situation is undesirable, but on the other, we can't manage resources if we can't probe an active node.
> 
> I'm thinking of reverting the node-pending-timeout portion of this PR. Alternatively maybe we could raise the default to something high (maybe 3h?).
> 
> Before node-pending-timeout, would it have been possible for someone to use a permanently Corosync-only node as a quorum node, or would it have blocked resource management?

AFAICT, a node that has only joined corosync membership is always marked as "pending".  And any "pending" node always blocks resource management. 

This PR didn't change that. The purpose of node-pending-timeout was to help users automatically get out of possibly unexpected forever "pending" situation.

> If so, I'd lean to reverting (at least until the next major-version release) to not break that use case. If not, a high default may be sufficient for other use cases (what comes to mind is someone having cluster trouble starting Corosync manually to make sure it comes up OK and maybe do some troubleshooting before starting Pacemaker manually).

Sounds like a reasonable use case.

Reply:
> FYI, I'd revert "add pending-node-timeout" through "trigger a new transition". 

We could revert them if we are not sure it's expected to automatically get out of "pending" situation.

> I think everything else would be useful to keep.

Okay.